8.1 support

Hyumiris opened the pull request.

Hyumiris committed:
naming for tests and fixed header only response

Hyumiris committed:
some simple testcases that already work

Hyumiris committed:
Merge commit 68f67347ba672d1f12661aa0ecd49252e8601582 into 8.1-support

Hyumiris committed:
reapply fix and use new flags syntax

Hyumiris committed:
remodeled incomingTraffic tests and improved headerCompleted

Hyumiris committed:
H2ReceivingState receive needs some work

(default pre and post handling)

Hyumiris committed:
new tests and minor change in testRequestResponseCommunication

added H2Header constants an H2StreamTests
new very long test for full 8.1 support (HCHHDDH)
removed first pushPromise test as this is not yet relevant
some frame creation helper in H2StreamTests

Hyumiris committed:
the basic functionality for 8.1

restructured ReceivingStates to act as a real StateMachine (transition on input)
distinction between general inputstream and dataInputStream in H2Stream

When the receiver has the correct class the receive method is called which in turn calls the subclassspecific process method. after that receive is used to save some state information and handle the completion of the communication if necissary.

Hyumiris committed:
minor fix

the stream didn't close after the final frame

Hyumiris committed:
renamed byteStreamIn to headerStreamIn

only headers are loaded into this stream as the data frames are loaded into a seperate stream in order to allow data frames be interspersed with push promise frames (forshadowing...)

Hyumiris committed:
Merge branch 'developer' into 8.1-support

Commit:
moved standard test data constructor to class side

Commit:
added some tests for class extensions

Hyumiris committed:
Merge branch '8.1-support' of https://github.com/hpi-swa-teaching/HTTP-2.git

Hyumiris committed:
Squeak 5.1 fixes

removed forks from the tests
removed addDeferredUIMessage from deliverResponse and closeOnError
some recategorization

ShirleyNekoDev commented on HTTP2-Core.package/H2Stream.class/instance/headerCompleted.st:
```diff
@@ -1,5 +1,9 @@
+	receivedHeader := (H2Header newFrom: self headerStreamIn contents in: self connection headerTableIn).
+	self headerStreamIn resetContents.
+	receivedHeader dictionary at: ':status' ifAbsent: [ ^self ]. "malformed request -> RFC 7540 8.1.2.4."
+	((receivedHeader at: ':status') < '200') ifTrue: [ ^self ]. "100 and 102 don't need handling, 101 does"
```
line 7/8 - should it be receivedHeader at: or ... dictionary at: ?

Hyumiris replied:
H2Header doesn't have ```at:ifAbsent```, only ```at:``` as a proxy for its dictionary.
I guess adding ```at:ifAbsent:``` to H2Header would be better. I'll add it.

Hyumiris committed:
removed "promise then" from testRequestResponseCommunication

Hyumiris committed:
Merge 'developer' into 8.1-support in prep for merge

Hyumiris committed:
fixed timeout error and remark from Salami555

LJBollmeier commented on HTTP2-Core.package/H2ContinuationFrameReceiver.class/instance/receiveContinuationFrame..st:
```diff
@@ -1,4 +1,5 @@
-	aFrame flags endHeaders ifTrue: [self transition]
\ No newline at end of file
+
+	"if endHeaders was not set"
```
Ist der Kommentar noch notwendig?

Hyumiris replied:
Sollte jemals jemand Error handling implementieren sollte an dieser Stelle die Überprüfung stattfinden ob die Flags des letzten Frames die folgenden Instruktionen zulassen oder ob ein Fehler geworfen werden sollte.

LJBollmeier commented on HTTP2-Core.package/H2ContinuationFrameReceiver.class/instance/receiveDataFrame..st:
```diff
@@ -0,0 +1,6 @@
+receiving
+receiveDataFrame: aFrame
+
+	"if endHeaders was set"
```
ebenso hier?

Hyumiris replied:
-||-

LJBollmeier commented on HTTP2-Tests.package/H2StreamTests.class/instance/testIncomingTraffic_HCC.st:
```diff
@@ -0,0 +1,21 @@
+testing
```
es wäre klug, wenn wir einfach einzelne sendemethoden für header/continutation...-frame machen, dann hat die testmethode den sinnvollen verlauf:
sendHeaderFrame
send ContinuationFrame
usw.

Hyumiris replied:
done

Hyumiris committed:
refactored the tests to look nicer and be clearer

frame generation expanded to also support endstream flags
setUp for stream and indexTable to streamline the tests
helper methods to immediately receive a frame split into any number of parts
